File: frontend/src/schemas/prediction.ts
/**
 * TypeScript Schemas - Frontend Contracts
 * 
 * These interfaces match the backend medical-grade schema exactly.
 * Strict typing prevents medical data mishandling.
 */

export type PredictionStatus = 'SUCCESS' | 'INCONCLUSIVE' | 'MODEL_ERROR' | 'INPUT_INVALID';
export type BinaryResult = 'BENIGN' | 'MALIGNANT';
export type StageResult = 'LOW' | 'MEDIUM' | 'HIGH';
export type RiskLevel = 'LOW' | 'MEDIUM' | 'HIGH' | 'INCONCLUSIVE';
export type ModelType = 'cnn_rnn' | 'vit';
export type ArtifactType = 'gradcam' | 'attention';

export interface ExplainabilityArtifact {
    id: number;
    prediction_id: number;
    artifact_type: ArtifactType;
    artifact_ref: string;
    expires_at: string;
    created_at: string;
}

export interface PredictionResponse {
    id: number;
    patient_id: number;
    model_id: number;

    // Status
    prediction_status: PredictionStatus;

    // Binary classification
    binary_result: BinaryResult | null;
    binary_confidence: number | null;

    // Stage classification
    stage_result: StageResult | null;
    stage_confidence: number | null;

    // Derived risk
    risk_level: RiskLevel;

    // Performance
    inference_time_ms: number;

    // Metadata
    created_at: string;

    // Explainability
    explainability_artifacts: ExplainabilityArtifact[];
}

export interface PredictionRequest {
    external_ref: string;
    file: File;
    model_type: ModelType;
}

export interface MetricsData {
    precision: number;
    recall: number;
    f1_score: number;
    confusion_matrix: {
        true_positive: number;
        true_negative: number;
        false_positive: number;
        false_negative: number;
    };
}

File: backend/app/schemas/prediction.py
"""
Prediction Schemas - Medical-Grade Design

Uses ENUMs from database models for type safety and consistency.
Explicit uncertainty states and confidence tracking.
"""

from pydantic import BaseModel, Field, field_validator
from datetime import datetime
from typing import Optional
from decimal import Decimal


# Import ENUMs from database models for consistency
from app.db.models import (
    PredictionStatus,
    BinaryResult,
    StageResult,
    RiskLevel,
    ArtifactType
)


# ============================================================================
# REQUEST SCHEMAS
# ============================================================================

class PredictionCreate(BaseModel):
    """
    Schema for creating a prediction
    
    This is used by the API endpoint to receive prediction data
    """
    patient_id: int
    model_id: int
    
    prediction_status: PredictionStatus
    risk_level: RiskLevel
    inference_time_ms: int
    
    # Binary classification (optional)
    binary_result: Optional[BinaryResult] = None
    binary_confidence: Optional[float] = Field(None, ge=0.0, le=1.0)
    
    # Stage/risk classification (optional)
    stage_result: Optional[StageResult] = None
    stage_confidence: Optional[float] = Field(None, ge=0.0, le=1.0)
    
    class Config:
        use_enum_values = True


# ============================================================================
# RESPONSE SCHEMAS
# ============================================================================

class PredictionSummary(BaseModel):
    """
    Summary view of a prediction (for lists)
    """
    id: int
    prediction_status: PredictionStatus
    risk_level: RiskLevel
    binary_result: Optional[BinaryResult] = None
    binary_confidence: Optional[Decimal] = None
    created_at: datetime
    
    class Config:
        from_attributes = True
        use_enum_values = True


class Prediction(BaseModel):
    """
    Full prediction response schema
    """
    id: int
    patient_id: int
    model_id: int
    
    # Status and results
    prediction_status: PredictionStatus
    binary_result: Optional[BinaryResult] = None
    binary_confidence: Optional[Decimal] = None
    stage_result: Optional[StageResult] = None
    stage_confidence: Optional[Decimal] = None
    risk_level: RiskLevel
    
    # Performance metrics
    inference_time_ms: int
    
    # Metadata
    created_at: datetime
    
    class Config:
        from_attributes = True
        use_enum_values = True


class PredictionWithExplainability(Prediction):
    """
    Prediction with explainability artifacts
    """
    explainability_artifacts: list['ExplainabilityArtifactResponse'] = []
    
    class Config:
        from_attributes = True
        use_enum_values = True


# ============================================================================
# EXPLAINABILITY SCHEMAS
# ============================================================================

class ExplainabilityArtifactCreate(BaseModel):
    """Schema for creating explainability artifact"""
    prediction_id: int
    artifact_type: ArtifactType
    artifact_ref: str
    expires_in_hours: int = 24


class ExplainabilityArtifactResponse(BaseModel):
    """Response schema for explainability artifact"""
    id: int
    prediction_id: int
    artifact_type: ArtifactType
    artifact_ref: str
    expires_at: datetime
    created_at: datetime
    
    class Config:
        from_attributes = True
        use_enum_values = True


# ============================================================================
# STATISTICS SCHEMAS
# ============================================================================

class PredictionStatistics(BaseModel):
    """Statistics about predictions"""
    total: int
    by_risk_level: dict[str, int]
    by_status: dict[str, int]


# Rebuild model to handle forward references
PredictionWithExplainability.model_rebuild()

